File: Challenge 8.py
'''
Write a program that will accept someone’s date of birth and
work out whether they can vote (i.e. Are they 18?)
'''
from datetime import datetime


today = datetime.now()
today = today.date()

date_of_birth = input('What is your date of birth? YYYY-MM-DD ')
dob = datetime.strptime(date_of_birth, '%Y-%m-%d')
dob = dob.date()

if (today.year - dob.year) < 18:
    print('Your are under age.')
elif (today.year - dob.year) == 18:
    if today.month < dob.month:
        print('You are under age.')
    elif today.month == dob.month:
        if today.day < dob.day:
            print('You are under age.')
        else:
            print('You are 18+.')
    else:
        print('You are 18+.')
else:
    print('You are 18+.')

File: Challenge 8.java
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.Scanner;
import java.util.concurrent.TimeUnit;

class VoteAgeValidator {

    public static String getTodaysDate() {
        LocalDateTime today = LocalDateTime.now();
        DateTimeFormatter f_today = DateTimeFormatter.ofPattern("dd/MM/yyyy");
        String date = today.format(f_today);
        return date;
    }

    public static String getDoB() {
        System.out.println("What is your date of birth?");
        System.out.println("Format: dd/mm/yyy");
        System.out.print(">> ");
        Scanner sc_input = new Scanner(System.in);
        String dob = sc_input.nextLine();
        return dob;
    }

    public static int getTimeDelta(String today, String dob) {
        double diff = 0;
        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyy");
        try {
            Date f_date = sdf.parse(today);
            Date s_date = sdf.parse(dob);
            long diffMillies = Math.abs(s_date.getTime() - f_date.getTime());
            diff = TimeUnit.DAYS.convert(diffMillies, TimeUnit.MILLISECONDS);
            // Just basic 365 year, no leap years in calculations included.
            // Not very exact but enough for exercise.
            diff = (diff / 365);
        } catch (ParseException e) {
            return 0;
        }
        return (int) diff;
    }
    
    public static void main(String[] args) {
        String dob = getDoB();
        String today = getTodaysDate();
        int result = getTimeDelta(today, dob);
        System.out.println(result);
        if (result < 18) {
            System.out.println("You are too young to vote.");
        } else {
            System.out.println("You can vote.");
        }
    }
}
